addon/array/fragment.js

```diff
@@ -81,15 +31,45 @@ const FragmentArray = StatefulArray.extend({
 
   options: null,
 
-  /**
-    @method _normalizeData
-    @private
-    @param {Object} data
-  */
-  _normalizeData(data) {
-    let content = this.content;
+  objectAt(index) {
+    const recordData = this._super(index);
+    if (recordData === undefined) {
+      return;
+    }
+    const internalModel = this.store._internalModelForResource(recordData.identifier);
+    if (internalModel === undefined) {
+      return;
+    }
+    return internalModel.getRecord();
+  },
+
+  _normalizeData(data, index) {
+    assert(
+      `You can only add '${this.modelName}' fragments or object literals to this property`,
+      typeOf(data) === 'object' || isInstanceOfType(this.store.modelFor(this.modelName), data)
+    );
+
+    if (isFragment(data)) {
+      const recordData = recordDataFor(data);
+      setFragmentOwner(data, this.recordData, this.key);
+      return recordData;
+    }
+    const existing = this.objectAt(index);
+    if (existing) {
+      existing.setProperties(data);
+      return recordDataFor(existing);
+    }
+    const fragment = this.store.createFragment(this.modelName, data);
+    setFragmentOwner(fragment, this.recordData, this.key);
+    return recordDataFor(fragment);
+  },
 
-    return normalizeFragmentArray(this, content, data, true);
+  replace(start, deleteCount, items) {
+    assert('The third argument to replace needs to be an array.', isArray(items));
+    const data = this.currentState.slice();
+    data.splice(start, deleteCount, ...items.map((item, i) => this._normalizeData(item, i)));
```

Comment:
```suggestion
    data.splice(start, deleteCount, ...items.map((item, i) => this._normalizeData(item, start + i)));
```

Reply:
without this, if I have a fragment array [{name: Obj1}], and I try to addObject({name: Obj2}), I get two objects with {name: Obj2} because you use the index of the item in the added items array instead of the index of the item in the actual array to look for existing, therefore going back to [0] and modifying that in normalizeData.

Reply:
Confirmed with a test - this only affects adding Objects, not full Fragments, since there's a return in there if the object you're adding is already a fragment. I have a PR with a test showing this. I'll create it.

Reply:
https://github.com/dwickern/ember-data.model-fragments/pull/2